README.md
# Running QuantaStor Ansible playbooks

## Step 1: install ansible
NOTE: best to do this on a stock ubuntu 16.04 VM as your Ansible controller else you may run into package pinning issues

    sudo apt-get update
    sudo apt-get install ansible git python3-pip
    pip3 install requests
    sudo python3 -m pip install quantastor-qsclient

## Step 2: clone quantastor ansible module git repository

    cd /path/to/dir
    git clone https://github.com/OSNEXUS/qsansible.git

## Step 3: tell ansible about one of your quantastor test boxes by editing /etc/ansible/hosts and adding a section like shown below:

    nano /etc/ansible/hosts

Modify this file with the following text (see playbooks/hosts.example):

    [qsservers]
    x.x.x.x qs_username='username' qs_password='password'


replace above IP with the IP of your QuantaStor appliance or one within a grid and the credantials with your own username and password.

## Step 3: copy the QuantaStor ansible module files into place, here I'm using a symbolic link to the source code in my git checkout rather than copying the content NOTE: installation location for 'qs_client.py' may differ on your system.

    cd /path/to/dir/qsansible
    ln -s /path/to/dir/qsansible/quantastor.py /usr/local/lib/python3.6/dist-packages/ansible/module_utils/.
    ln -s /usr/local/lib/python3.6/dist-packages/quantastor/qs_client.py /usr/local/lib/python3.6/dist-packages/ansible/module_utils/.
    ln -s /path/to/dir/qsansible/quantastor/ /usr/local/lib/python3.6/dist-packages/ansible/modules/storage/

## Step 4: run a playbook to test the module

You can find example playbooks in the `/path/to/dir/qsansible/playbooks/` location. If you are in the root of the git repo you can run the following commands:

    cd playbooks/
    ansible-playbook qstest_addhost.yml -vvv

## Step 5: verification.. make sure that the above commands ran successfully


## MISC NOTES: these things are not needed but can be used to get a newer copy of ansible from github and then puts it into your python path.  If you do this you don't need to apt-get install ansible
    sudo apt-add-repository ppa:ansible/ansible
    git clone https://github.com/ansible/ansible.git
    export PYTHONPATH="/root/ansible:/root/ansible/lib"
    python ./ansible/lib/ansible/modules/storage/quantastor/quantastor_host.py 

# Testing using ansible playbooks

This section will detail how to set up a test to run the 'qstest_addshare_test.yml' playbook. This playbook removes the share 'testShare'
if it exists, then creates a new share called 'testShare' with various parameters set/assigned. The purpose of these instructions are to insure that 
all of the prerequisits (groups, users, pools, etc...) to run this testing-playbook are set up such that the instructions are executed successfully.

Before running this playbook, you should at least follow steps 1 - 3 in the section above.

## Step 1: Create a user-group named 'testGroup' using the QuantaStor CLI, REST service, or UI.

## Step 2: Create a storage pool named 'DefaultPool' using the QuantaStor CLI, REST service, or UI.

## Step 3: Navigate to the playbook directory. 

Playbooks can be found in the qsansible git repository `./path/to/dir/qsansible/playbooks/`

## Step 4: Now you have all of the prerequisits neeeded. Run the playbook in verbose mode using the following command:

    ansible-playbook <playbook>.yml -vvv

eg.
    ansible-playbook qstest_addshare.yml -vvv

Result: If you look at your target QuantaStor_host UI you should see that a Network Share named 'testShare' 
has been created with the parameters listed in the playbook set. Alternatively you can run one of these commands to see that you new share exists:

### CLI command from a QuantaStor box on your system :
    qs share-list

### Using qstorapi :
    curl -u admin:password "https://[hostIP]:8153/qstorapi/networkShareGet?networkShare=testShare" -k

To setup your own playbook see documentation in ansible modules. (quantastor_host.py, quantastor_share.py, quantastor_volume.py)

release/release_notes.txt
Release Notes Document for QuantaStor-Ansible Modules

qsansible v.1.0 - compatable with QuantaStor 5.5.0 [Date: 12/19/2019]

  >  Error message changed from "To create a normal share, the 'parent' parameter must be a valid storage pool." -> "To create a normal share, the 'pool' parameter must be a valid storage pool." [ QSTOR-6629 ]
  >  Host remove operation now allows you to pass the 'flags' parameter allowing you to delete a host that is a member of a host group by passing the force flag. [ QSTOR-6631 ]
  >  Published Ansible Module and Python client library that use QuantaStor REST API for automation. [ QSTOR-6025 ]
  >  Running the playbooks to unassign a volume from a host that doesn't exist no longer throws an error. Instead the unassignment task exits with changed=false. [ QSTOR-6630 ]


quantastor/quantastor_host.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# (c) 2018, OSNEXUS Engineering (eng@osnexus.com), Steven Umbehocker (steve@osnexus.com)
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = r'''
---
module: quantastor_host
version_added: '4.6'
short_description: Manage iSCSI/FC host entries within QuantaStor storage grids
description:
- Create, delete or modify hosts within a QuantaStor storage grid.
author:
- Steven Umbehocker
options:
  host:
    description:
    - The hostname of the host.
    required: true
  state:
    description:
    - Host with an associated IQN(s) and/or WWPN(s) must be added before Storage Volume may be assigned to a given Host.
    - Removing a Host will disconnect all volumes assigned to it.  Hosts may be grouped into Host Groups.
    default: present
    choices: [ absent, present ]
  initiators:
    description:
    - List of iSCSI IQNs and/or FC WWPNs for the Host.
extends_documentation_fragment:
- quantastor
'''

EXAMPLES = r'''
- name: Create new Host entry called "foo" in the storage grid
  quantastor_host:
    quantastor_hostname: 10.10.10.2
    quantastor_username: admin
    quantastor_password: password
    host: foo
    initiators: iqn.foo

- name: Delete Host entry called "foo" from the storage grid
  quantastor_host:
    quantastor_hostname: 10.10.10.2
    quantastor_username: admin
    quantastor_password: password
    host: foo
    state: absent

- name: Create a Host called "bar" with WWPN and IQN entries in the storage grid
  quantastor_host:
    quantastor_hostname: 10.10.10.2
    quantastor_username: admin
    quantastor_password: password
    host: bar
    initiators:
    - 00:00:00:00:00:00:00
    - 11:11:11:11:11:11:11
    - iqn.1994-05.com.redhat:12345678

- name: Map Host "foo" to Storage Volume "volume2"
  quantastor_host:
    quantastor_hostname: 10.10.10.2
    quantastor_username: admin
    quantastor_password: password
    host: foo
    volume: volume2
'''

RETURN = r'''
'''

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.quantastor import quantastor_argument_spec
from ansible.module_utils.qs_client import quantastor_sdk_enabled
from ansible.module_utils.qs_client import QuantastorClient
from ansible.module_utils.qs_client import Host

# Helper function forms 2 sets from given arguments and _initiatorPortList then returns 
# the Difference-set or the Intersection-set based on the given mode (true: intersection, false: difference).
def getIntersectionDifference(argset, hostset, mode):
    ret = set()
    tempArgs = set()
    tempHosts = set()

    for arg in argset:
        tempArgs.add(arg)
    for init in hostset:
        tempHosts.add(init['name'])

    if mode:
        ret = tempArgs.intersection(tempHosts)
    else:
        ret = tempArgs.difference(tempHosts)

    return ret


def main():
    argument_spec = quantastor_argument_spec()
    argument_spec.update(dict(
        host=dict(type='str'),
        hosts=dict(type='list'),
        hostgroup=dict(type='str'),
        description=dict(type='str'),
        initiators=dict(type='list'),
        volume=dict(type='str'),
        state=dict(type='str', default='present', choices=['absent', 'present']),
        flags=dict(type='str',default='0'),
    ))

    # System checks
    module = AnsibleModule(argument_spec, supports_check_mode=True)
    if not quantastor_sdk_enabled():
        module.fail_json(msg='QuantaStor python SDK is required for this module.')

    client = QuantastorClient.from_module(module)
    try:
        client.storage_system_get()
    except:
        module.fail_json(msg='Unable to gather QuantaStor system information.')

    # Operation Variables
    state = module.params['state']    
    
    # Object Variables
    host = None
    hostgroup = None

    # Bailout checks
    if module.params['host'] and module.params['hostgroup']:
        module.fail_json(msg="Cannot perform operation that has both the 'host' and 'hostgroup' parameters set.")
    if not (module.params['host'] or module.params['hostgroup']):
        module.fail_json(msg="Cannot perform operation unless one of either the 'host' or 'hostgroup' parameters are specified.")
    if module.params['volume'] and module.params['initiators']:
        module.fail_json(msg="Invalid argument specification. You cannot specify both the 'volume' and the 'initiators' arguments together.")

    if state == "present":
        # create host or add host initiators
        if module.params['host']:
            try:
                host = client.host_get(host=module.params['host'])
                hostId = module.params['host']
            except Exception as e:
                # case: Add new host entry
                if module.params['initiators']:
                    pass
                else:
                    module.fail_json(msg="Operation failed because target host '%s' does not exist. Error was '%s'." % (module.params['host'],str(e)))
        
        # create hostgroup
        if module.params['hostgroup']:
            try:
                task, hostgroup = client.host_group_get(hostGroup=module.params['hostgroup'])
                # case: Trying to create a hostgroup with name that already exists.
                if module.params['hosts']:
                    module.exit_json(changed=False)
                hostId = module.params['hostgroup']
            except Exception as e:
                # case: Create new hostgroup
                if module.params['hosts']:
                    pass
                else:
                    module.fail_json(msg="Operation failed because target hostgroup '%s' does not exist. Error was '%s'." % (module.params['hostgroup'],str(e)))

        # attach volume to host or hostgroup
        if module.params['volume']:   
            try:
                client.storage_volume_get(module.params['volume'])
            except Exception as e:
                module.fail_json(msg="To assign a storage volume to a host/hostgroup, the target volume '%s' must exist. Error was '%s'." % (module.params['volume'],str(e)))

            try:
                client.storage_volume_acl_get(storageVolume=module.params['volume'],host=hostId)
                module.exit_json(changed=False)
            except Exception as e:
                # case: ACL add between module.params['volume'] and a host/hostgroup
                if (host or hostgroup):
                    pass
                else:
                    module.fail_json(msg="Cannot attach storage volume '%s' to a host or host group that does not exist. Error was '%s'." % (module.params['volume'], str(e)))

    if state == "absent":
        # delete host or remove host initiators
        if module.params['host']:
            try:
                host = client.host_get(host=module.params['host'])
                hostId = module.params['host']
            except Exception as e:
                if not (module.params['initiators'] or module.params['volume']):
                    module.exit_json(changed=False)
                else:
                    module.fail_json(msg="Cannot remove host entry '%s' because it does not exist. Error was '%s'" % (module.params['host'],str(e)))
        
        # delete hostgroup
        if module.params['hostgroup']:
            try:
                task, hostgroup = client.host_group_get(hostGroup=module.params['hostgroup'])
                hostId = module.params['hostgroup']
            except Exception as e:
                if not (module.params['volume']):
                    module.exit_json(changed=False)
                else:
                    module.fail_json(msg="Cannot remove hostgroup '%s' because it does not exist. Error was '%s'" % (module.params['hostgroup'],str(e)))
        
        # detach storage volumes from host or hostgroup
        if module.params['volume']:
            try:
                client.storage_volume_get(storageVolume=module.params['volume'])
            except Exception as e:
                module.fail_json(msg="Failed to detatch storage volume ACL because storage volume '%s' does not exist. Error was '%s'." % (module.params['volume'],str(e)))

            try:
                client.storage_volume_acl_get(storageVolume=module.params['volume'],host=hostId)
            except Exception as e:
                module.exit_json(changed=False)



    # Create/Add Operations
    if state == 'present':
        #CREATE HOST GROUP
        if not hostgroup and module.params['hosts']:
            try:
                client.host_group_create(
                    name=module.params['hostgroup'],
                    description=module.params['description'],
                    hostList=module.params['hosts'],
                    flags=module.params['flags']
                    )
                module.exit_json(changed=True)
            except Exception as e: 
                module.fail_json(msg="Failed to create host group '%s' with hosts '%s', error was '%s'." % (module.params['hostgroup'], ','.join(module.params['hosts']), str(e)))
            
        #ASSIGN VOLUME TO HOST OR HOSTGROUP
        elif (module.params['hostgroup'] or module.params['host']) and module.params['volume']:
            try:
                client.storage_volume_acl_add_remove_ex(
                    storageVolumeList=module.params['volume'], 
                    host=hostId,
                    modType=0, #OSN_CMN_MOD_OP_ADD
                    flags=module.params['flags']
                    )
                module.exit_json(changed=True)
            except Exception as e: 
                module.fail_json(msg="Failed to create new volume assignment entry to host '%s/%s', error was '%s'." % (module.params['volume'], module.params['host'], str(e)))

        #ADD HOST ENTRY
        elif not host and module.params['initiators'] and module.params['host']:
            try:
                client.host_add(
                        hostname=module.params['host'], 
                        iqn=module.params['initiators'][0],
                        description=module.params['description'],
                        flags=module.params['flags']
                        )
                initiators = iter(module.params['initiators'])
                next(initiators)
                for port in initiators:
                    try:
                        client.host_initiator_add(
                            host=module.params['host'], 
                            iqn=port,
                            flags=module.params['flags']
                            )
                    except Exception as e: 
                        module.fail_json(msg="Failed to create new host initiator entry '%s', error was '%s'." % (port, str(e)))
                module.exit_json(changed=True)
            except Exception as e: 
                module.fail_json(msg="Failed to create new host entry '%s', error was '%s'." % (module.params['host'], str(e)))
            module.fail_json(msg="Failed to create new host entry '%s'." % (module.params['host']))

        #ADD HOST INITIATORS TO HOST ENTRY
        elif host and module.params['initiators']:
            missingInitiators = getIntersectionDifference(module.params['initiators'],host._initiatorPortList,False) #return difference
            if len(missingInitiators) > 0:
                for port in missingInitiators:
                    try:
                        client.host_initiator_add(host=module.params['host'], iqn=port)
                    except Exception as e: 
                        module.fail_json(msg="Failed to create new host initiator entry '%s', error was '%s'." % (port, str(e)))
                module.exit_json(changed=True)
            else:
                module.exit_json(changed=False)

    # Delete/Remove operations
    elif state == 'absent':
        #DELETE HOST GROUP
        if hostgroup and not module.params['volume']:
            try:
                client.host_group_delete(hostGroup=module.params['hostgroup'])
                module.exit_json(changed=True)
            except Exception as e: 
                module.fail_json(msg="Failed to delete host group '%s', error was '%s'." % (module.params['hostgroup'], str(e)))
            module.fail_json(msg="Failed to delete host group '%s'." % (module.params['hostgroup']))

        #UNASSIGN VOLUME FROM HOST or HOSTGROUP 
        elif module.params['volume']:
            try:
                client.storage_volume_acl_add_remove_ex(
                    storageVolumeList=module.params['volume'], 
                    host=hostId,
                    modType=1
                    )
                module.exit_json(changed=True)
            except Exception as e: 
                module.fail_json(msg="Failed to remove volume assignment entry for host (or hostgroup) '%s/%s', error was '%s'." % (module.params['volume'], hostId, str(e)))
            module.fail_json(msg="Failed to remove volume assignment entry for volume '%s'." % (module.params['volume']))

        #REMOVE HOST
        elif host and not module.params['volume'] and not module.params['initiators']:
            try:
                client.host_remove(module.params['host'])
                module.exit_json(changed=True)
            except Exception as e: 
                module.fail_json(msg="Failed to remove host entry '%s', error was '%s'." % (module.params['host'], str(e)))
            module.fail_json(msg="Failed to remove host entry '%s'." % (module.params['host']))

        #REMOVE HOST INITIATORS FROM HOST
        elif host and module.params['initiators']:
            removeInitiators = getIntersectionDifference(module.params['initiators'],host._initiatorPortList,True) # return intersection
            if len(removeInitiators) > 0:
                for port in removeInitiators:
                    try:
                        client.host_initiator_remove(module.params['host'], port)
                    except Exception as e: 
                        module.fail_json(msg="Failed to remove host initiator entry '%s', error was '%s'." % (port, str(e)))
                module.exit_json(changed=True)
            else:
                module.exit_json(changed=False)

    else:
        module.fail_json(msg="Invalid state '%s', must be either 'present' or 'absent'." % module.params['state'])

    module.exit_json(changed=False)


if __name__ == '__main__':
    main()


quantastor/quantastor_volume.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# (c) 2019, OSNEXUS Engineering (eng@osnexus.com), Steven Umbehocker (steve@osnexus.com)
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = r'''
---
module: quantastor_volume
version_added: '4.6'
short_description: Manage Storage Volume (iSCSI/FC) block storage operations within QuantaStor storage grids
description:
- Create, delete or modify storage volumes within a QuantaStor storage grid.
author:
- Steven Umbehocker, Seth Cagampang
options:
  volume:
    description:
    - Name of the new storage volume to be created (of any type volume, snapshot, clone, etc).
  volumeType:
    description:
    - Specify the volume type to be created (*normal, snapshot, clone)
  size:
    description:
    - size of the volume to be created in MB
  state:
    description:
    - Creates (present) or deletes (absent) a storage volume
    default: present
    choices: [ absent, present ]

Volume-Delete options:
  deleteChildren:
    description:
    - Set to 'true' to recursively delete child snapshots.

volumeType = snapshot Options:
  volume:
    description:
    - Optional name for storage volume snapshot. If not specified, an auto-generated name will be assigned to the snapshot.
  parent:
    description:
    - Specify the parent volume for this snapshot.
  description:
    description:
    - Gives a description to the snapshot
  accessMode:
    description:
    - Sets the access mode for the storage volume snapshot. [none, readonly, *readwrite]
  count:
    description:
    - Sets the number of snapshots of the specified volume to create.
  
  flags:
    description:
    - Optional flags for the operation.
extends_documentation_fragment:
- quantastor
'''

EXAMPLES = r'''
- name: Create new Storage Volume of size 1GB called volumeA in pool DefaultPool
  quantastor_volume:
    quantastor_hostname: 10.10.10.2
    quantastor_username: admin
    quantastor_password: password
    volume: volumeA
    pool: DefaultPool
    size: 1024

- name: Create a snapshot of volumeA called snapA
  quantastor_volume: 
    quantastor_hostname: 10.10.10.2
    quantastor_username: admin
    quantastor_password: password 
    parent: volumeA
    volume: snapA
    description: desc-snapA
    accessMode: 0
    count: 1
    flags: 0

- name: Delete Storage Volume snapshot named snapA of volumeA
  quantastor_volume:
    quantastor_hostname: 10.10.10.2
    quantastor_username: admin
    quantastor_password: password
    volume: snapA
    state: absent

- name: Delete Storage Volume named volumeA
  quantastor_volume:
    quantastor_hostname: 10.10.10.2
    quantastor_username: admin
    quantastor_password: password
    volume: volumeA
    state: absent
'''

RETURN = r'''
'''

from ansible.module_utils.basic import AnsibleModule
from os import environ
import requests
from requests.auth import HTTPBasicAuth
from ansible.module_utils.quantastor import quantastor_argument_spec
from ansible.module_utils.qs_client import quantastor_sdk_enabled
from ansible.module_utils.qs_client import QuantastorClient

def main():
    argument_spec = quantastor_argument_spec()
    argument_spec.update(dict(
        volume=dict(type='str'),
        parent=dict(type='str'),
        pool=dict(type='str'),
        size=dict(type='str'),
        description=dict(type='str'),
        accessMode=dict(type='str'),
        count=dict(type='str'),
        state=dict(type='str', default='present', choices=['absent', 'present']),
        volumeType=dict(type='str', default='normal', choices=['normal','snapshot','clone']),
        #delete option
        deleteChildren=dict(type='bool', default=False),
        flags=dict(type='int', default=0),
    ))

    # System checks
    module = AnsibleModule(argument_spec, supports_check_mode=True)
    if not quantastor_sdk_enabled():
        module.fail_json(msg='QuantaStor python SDK is required for this module.')

    client = QuantastorClient.from_module(module)
    try:
        client.storage_system_get("")
    except:
        module.fail_json(msg='Unable to gather QuantaStor system information.')

    # Operational Variables
    state = module.params['state']
    volumeType = module.params['volumeType']

    # Bailout checks
    # all non-'snapshot' volume types require 'volume' parameter.
    if not module.params['volume'] and not volumeType == 'snapshot':
        module.fail_json(msg="To create/delete a '%s', the 'volume' parameter must be specified." % (module.params['volumeType']))
    
    if module.params['volume']:
        try:
            client.storage_volume_get(storageVolume=module.params['volume'])
            if state == 'present':
                # If you try to create (present), exit if a volume with that name already exists.
                module.exit_json(changed=False)
        except Exception as e:
            if state == 'absent':
                # If you try to delete (absent), exit if no volume with that name exists.
                module.exit_json(changed=False)
            else:
                # case: storage volume named: module.params['volume'] does not already exist and state = 'present'
                pass

    if not module.params['parent']:
        if not volumeType == 'normal':
            # all non-'normal' volume types require 'parent' parameter.
            module.fail_json(msg="To create a '%s', the 'parent' parameter must be specified." % (module.params['volumeType']))

    if volumeType == 'normal':
        if not module.params['size'] and not state == 'absent':
            # all normal shares need to have a 'size' argument
            module.fail_json(msg="To create a normal share you must provide a 'size' parameter.")
    else:
        try:
            client.storage_volume_get(storageVolume=module.params['parent'])
        except Exception as e:
            # all non-'normal' volume types require a vaild parent object
            module.fail_json(msg="To create a '%s', the 'parent' parameter must be a valid storage volume." % (module.params['volumeType']))

    
    # For create operations:
    if state == 'present':
        # normal volumes require the 'pool' parameter to be a valid storage pool.
        if not module.params['pool']:
            if volumeType == 'normal':
                module.fail_json(msg="To create a normal volume, the 'pool' parameter must be specified.")
        else:
            try:
                client.storage_pool_get(module.params['pool'])
            except:
                module.fail_json(msg="To create a normal volume, the 'pool' parameter must be a valid storage pool.")

    #Parameter interpretation
    deleteChildren = False
    if module.params['deleteChildren']:
        deleteChildren = module.params['deleteChildren']

    # Case: set recursive flag
    if deleteChildren:
        flags = 262144
    else:
        flags = module.params['flags']
    
    #CREATE
    if state == 'present':
        #NORMAL VOLUME
        if volumeType == 'normal':
            try:
                client.storage_volume_create_ex(
                            name=module.params['volume'],
                            size=module.params['size'], 
                            description=module.params['description'], 
                            provisionableId=module.params['pool']
                            )
            except Exception as e: 
                module.fail_json(msg="Failed to create storage volume '%s', error was '%s'." % (module.params['volume'], str(e)))

        #SNAPSHOT
        elif volumeType == 'snapshot':
            try:
                client.storage_volume_snapshot(
                    storageVolume=module.params['parent'],
                    snapshotName=module.params['volume'],
                    description=module.params['description'],
                    accessMode=module.params['accessMode'],
                    count=module.params['count'],
                    flags=module.params['flags'],
                    )
            except Exception as e:
                module.fail_json(msg="Failed to create snapshot '%s' for Storage Volume  '%s', error was '%s'." % (module.params['snapName'],module.params['volume'], str(e)))

        #CLONE
        elif volumeType == 'clone':
            #unimplemented
            module.exit_json(changed=False)

        module.exit_json(changed=True)

    #DELETE
    elif state == 'absent':
        try:
            client.storage_volume_delete(storageVolumeList=module.params['volume'],flags=flags)
        except Exception as e: 
            module.fail_json(msg="Failed to delete storage volume '%s', error was '%s'." % (module.params['volume'], str(e)))
        
        # Verify Deletion
        module.exit_json(changed=True)
    
    module.exit_json(changed=False)


if __name__ == '__main__':
    main()
